### PR TITLE
feat(react): Reactification of Settings and Profile Cont...

### LINKED ISSUE #15917: Translations are no longer working
I can't seem to find any page on beta that has a Spanish translation.

### LINKED ISSUE #16237: Add isNewDataVisCert property
#### Issue Description
On the user object, we will need both a `isDataVisCert` and an `isNewDataVisCert` property.

`isDataVisCert` should refer to what we will call the "Legacy Data Visualization Certificate" and is `isNewDataVisCert` should refer to the "Data Visualization Certificate".

### LINKED ISSUE #16270: MY PROFILE: Previously Completed Challenges showing as NA
#### Challenge Name

My Profile -> Show me my public profile

#### Issue Description

Some previously completed challenged are not loading correctly.
They are listed in the profile as 'NA'
Screenshot is below.

#### Browser Information

* Browser Name, Version:  
* Operating System: FireFox 57.0 (64-bit)  and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
* Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-bit 

#### Your Code

```js



```
#### Screenshot

image

### LINKED ISSUE #16271: Social Media Buttons - Unable to remove account(s) - account not found
#### Challenge Name

My Profile -> Update my Settings 
Add my twitter account and add my linked in account

#### Issue Description

added my twitter account to my profile. when trying to remove the account, it gives an error saying no account found. I have not tried for linked in.

#### Browser Information

* Browser Name, Version:  
* Operating System: FireFox 57.0 (64-bit)  and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
* Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-

#### Your Code

```js



```
#### Screenshot

image

image

### LINKED ISSUE #16273: Social Media: Adding Twitter or LinkedIn modifies the other mned
#### Challenge Name

My Profile -> Update my Settings
Add my twitter account and add my linked in account

#### Issue Description

After adding only the twitter account, the linkedin account becomes active as if a linkedin account was added but was not.
cannot remove the falsely account. URI does not match.
but the twitter logo is available on public profile with privacy settings disabled.

#### Browser Information

* Browser Name, Version:  
* Operating System: FireFox 57.0 (64-bit)  and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
* Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-

#### Your Code

```js



```
#### Screenshot

image

image

image

### LINKED ISSUE #16278: Available For Hire does not update the public profile if they are/not available
#### Challenge Name

Profile - Update My Settings - Available for Hire

#### Issue Description

When enabling the 'Available for Hire' it does not appear in the public profile 
that the camper is available for hire.
Maybe add a "Available For Hire: Yes" or "Available For Hire: Yes" depending on the camper's selection
Maybe just above the brownie/cookie points value

#### Browser Information

* Browser Name, Version: 
* Operating System: 
* Mobile, Desktop, or Tablet: 

#### Your Code

```js



```
#### Screenshot

image

### LINKED ISSUE #16279: Language Option menu does not close and turns green when mouse is moved away
#### Challenge Name

My Profile - Update My Settings - Display Challenges in

#### Issue Description

When leaving the dropdown menu open and move the mouse away from the drop down
the menu does not close and it turns green

#### Browser Information

* Browser Name, Version:  
* Operating System: FireFox 57.0 (64-bit)  and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
* Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-bit

#### Your Code

```js



```
#### Screenshot

image

### LINKED ISSUE #16283: Public Profile - Social Media Buttons look and behave differently
#### Challenge Name
https://beta.freecodecamp.org/<username>
Privacy Settings must be turned off and a social media account linked to verify

GitHub Button: By Default is Green instead of white and turns a different shade of light grey on hover
Twitter Button: By Default is White and turns a different shade of dark green (not the FCC green) on hover
LinkedIn: unknown since there is an issue in an inability to remove social accounts, due to privacy not linking my linkedin. Please verify

For UX consistency, the buttons should appear and behave the same.

#### Browser Information

* Browser Name, Version:  
* Operating System: FireFox 57.0 (64-bit)  and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
* Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-bit

#### Your Code

```js



```
#### Screenshot

image

image

image

### LINKED ISSUE #16285: Changes to privacy setting does not show popup confirm changes
#### Challenge Name

My Profile -> Update my Settings -> Privacy Settings

#### Issue Description

When toggling privacy settings between On or Off, there is no confirmation the changes were done. When you verify your public profile the changes do take effect.

Production FCC does alert when the changes are executed.
The screenshot below is from Production FCC. 
#### Browser Information

Browser Name, Version:
Operating System: FireFox 57.0 (64-bit) and Chrome Version 63.0.3239.84 (Official Build) (64-bit)
Mobile, Desktop, or Tablet: Laptop Windows 10 Professional 64-bit

#### Your Code

```js



```
#### Screenshot

Production FCC:
image

### LINKED ISSUE #16587: [Beta] In Firefox setting page can't align properly with navigation bar.
This alignment problem happen only in setting page. Its happen when browser zoom set in more than 40% and current monitor resolution set in 1920 * 1080 (16:9). But in 40% zoom setting page is harder to read. For more details please check below screenshots. 
### Set in 40% zoom
40 _zoom
### Set in 100% zoom
100 _zoom

Current Environment: 
* Browser Name, Version:  Firefox 58
* Operating System: Linux Mint 18.3
* Mobile, Desktop, or Tablet: Desktop with 21.5" Monitor

### LINKED ISSUE #16597: Resequence bio edit fields
cursor_and_settings___freecodecamp

### LINKED ISSUE #16598: Report user button
As a user, if I'm authenticated and viewing someone else's profile, I should be able to report them by clicking a "report user profile" button.

### LINKED ISSUE #16599: Update email settings
cursor_and_settings___freecodecamp

### LINKED ISSUE #16600: Remove "available for hire" section of settings

### PR DESCRIPTION
**Pull request by Bouncey**
The cliff notes on this PR:

- Consolidate the majority of  `updateMy***` actions into one `updateUserBackend` action
> Berkeley's third rule of Redux: "Never use two actions where one will do."
- Add more options to `/settings` to allow the user a more control over their public profile
- Consolidate all `verifyCert` endpoints into one `/certificate/verify` endpoint
- Consolidate all `showCert` endpoints into one `/c/:username/:cert`endoint
- Allow the user to complete projects via the challenge itself or the settings page, either of which updates `user.challengeMap` as well as `user.projects`.
- Allow the user to claim and see certs from `/settings`
- Allow the user to update their username from `/settings`, with verification from the backend that the username is available
- Fixes an issue where `lang` is `undefined` in the `mapService`
- Fixes an issue where the settings container wraps itself off screen in FireFox
- Fixes styling of flash messages

Their might be other things but I cannot remember right now.

closes #15917 
closes #16237
closes #16270 
closes #16271 
closes #16273 
closes #16278 
closes #16279 
closes #16283 
closes #16285 
closes #16587 
closes #16597 
closes #16598
closes #16599 
closes #16600

### CODE REVIEW

**Review by Speca92 — APPROVED** [review]

File: server/boot/user.js [hunk 1]
```diff
@@ -11,24 +9,18 @@ import {
   frontEndLibsId,
   jsAlgoDataStructId,
   dataVisId,
+  dataViz2018Id,
```
Comment (raisedadead) [inline server/boot/user.js]: `dataViz2018Id` may be cleaner calling it as `dataVizModernId`

Prefer, doing this, here and everywhere?

Comment (Bouncey) [reply server/boot/user.js]: My thinking was, we can overhaul the curriculum again in three years time, what would be call it then?

So I though a yearly name-space would help any further expansion in years to come

Comment (raisedadead) [reply server/boot/user.js]: Hmm.. seems reasonable. Thanks, I'll pull down an check soon.

File: common/app/routes/Profile/components/Certificates.jsx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+  currentCerts,
+  legacyCerts,
+  hasLegacyCert,
+  hasModernCert,
+  username
+}) {
+  const renderCertShowWithUsername = _.curry(renderCertShow)(username);
+  return (
+    <div>
+      <h2 className='text-center'>FreeCodeCamp Certificates</h2>
```
Comment (raisedadead) [inline common/app/routes/Profile/components/Certificates.jsx]: FreeCodeCamp => freeCodeCamp

**Review by raisedadead — DISMISSED** [review]

File: common/app/routes/Profile/components/Certificates.jsx (hunk 2 above)
Comment (QuincyLarson) [reply common/app/routes/Profile/components/Certificates.jsx]: @Bouncey has made this change.

**Comment by QuincyLarson** [comment]
Notes from QA session:

cursor_and_rm_-rf_node_modules____npm_install____node_seed____gulp_ _node_ _gulp_term_program_apple_terminal_ _159x45

cursor_and_settings___freecodecamp

cursor_and_profile___freecodecamp

**Comment by Bouncey** [comment]
about

email

internetpresence

portfolio

certs

timeline

reset

delete

**Comment by QuincyLarson** [comment]
@Bouncey From a functionality standpoint, these changes all look great. Thanks for creating the gifs, too! They make QA'ing much easier! LGTM.

@BerkeleyTrue what do you think - is this ready to merge?

File: common/app/redux/fetch-user-epic.js [hunk 3]
```diff
@@ -17,3 +22,21 @@ export default function getUserEpic(actions, _, { services }) {
         .catch(createErrorObservable);
     });
 }
+
+function getOtherUserEpic(actions$) {
```
Comment (Bouncey) [inline common/app/redux/fetch-user-epic.js]: @BerkeleyTrue

File: common/models/user.js [hunk 4]
```diff
@@ -510,69 +595,259 @@ module.exports = function(User) {
...
+        .do(() => {
+          this.username = newUsername;
+        })
+        .map(() => dedent`
+        Username updated successfully
+        `);
+    });
   };
 
+  User.getPublicProfile = function getPublicProfile(username, cb) {
```
Comment (Bouncey) [inline common/models/user.js]: @BerkeleyTrue

File: server/utils/publicUserProps.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  'points',
+  'portfolio',
+  'projects',
+  'streak',
+  'twitter',
+  'username',
+  'website'
+];
+
+export const userPropsForSession = [
```
Comment (Bouncey) [inline server/utils/publicUserProps.js]: @BerkeleyTrue

File: server/services/user.js [hunk 6]
```diff
@@ -1,72 +1,63 @@
...
+            ...progress,
+            challengeMap
           }))
-      )
+          .map(
+            user => ({
+              entities: {
+                user: {
+                  [user.username]: {
+                    ..._.pick(user, userPropsForSession),
```
Comment (Bouncey) [inline server/services/user.js]: @BerkeleyTrue

**Review by Bouncey — COMMENTED** [review]
@BerkeleyTrue I have tagged you in most of the relevant places re the new `getPublicProfile` remote method

File: server/services/user.js [hunk 7]
```diff
@@ -1,72 +1,63 @@
...
+  normaliseUserFields,
+  getProgress
+} from '../utils/publicUserProps';
 
 export default function userServices() {
   return {
     name: 'user',
-    read: (req, resource, params, config, cb) => {
-      const { user } = req;
+    read: async function readUserService(
```
Comment (BerkeleyTrue) [inline server/services/user.js]: Why is this an async func?

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by BerkeleyTrue** [comment]
@Bouncey Why is this coupled with the settings page? The settings page is already react.

File: common/app/Flash/redux/get-messages-epic.js [hunk 8]
```diff
@@ -3,13 +3,14 @@ import { ofType } from 'redux-epic';
...
-  fetchMessagesError
+  fetchMessagesError,
+  types
 } from './';
 import { types as app } from '../../redux';
 import { getJSON$ } from '../../../utils/ajax-stream.js';
 
 export default function getMessagesEpic(actions) {
-  return actions::ofType(app.appMounted)
+  return actions::ofType(app.appMounted, types.fetchMessages.start)
```
Comment (BerkeleyTrue) [inline common/app/Flash/redux/get-messages-epic.js]: I can't find where this change is used.

File: server/services/user.js (hunk 7 above)
Comment (Bouncey) [reply server/services/user.js]: That is left over from when I used `User.find$` to find a third party profile. It can be removed

**Comment by Bouncey** [comment]
I have a commit to remove the `async` and `fetchMessages` action. Both are leftover from previous itterations

The settings page has been heavily modified to add new options (mainly stuff that GitHub took care of) and remove depricated options, like social account linking, from the settings page. The functionality on the backend is still there, it should be removed in a subsequent PR.

**Comment by raisedadead** [comment]
> The functionality on the backend is still there, it should be removed in a subsequent PR. 

Can you just list the APIs or files? I'll take care of auditing and making a PR post this is merged?

**Comment by jonathan-grah** [comment]
@Bouncey, great work on this! From the GIFs, this looks great :smiley: I will pull this down when ready to properly QA everything.

**Pull request closed without merging**